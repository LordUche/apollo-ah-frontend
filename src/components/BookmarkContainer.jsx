/* eslint-disable react/jsx-indent */
/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-wrap-multilines */
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import BookmarkItem from '../views/BookmarkItem';
import { getBookmarkedArticles } from '../actions/getBookmark';
import * as selectors from '../selectors/getBookmarkSelector';

/**
 * @description - returns all articles bookmarked by a user
 * @return {JSX}
 */
function BookmarkContainer(props) {
  const [viewType, setViewType] = useState('row view-row');

  const { articles } = props;

  const getArticlesHandler = () => {
    props.getBookmarkedArticles();
  };

  const changeViewToHorizontal = () => {
    setViewType('row view-row list-view');
  };

  const changeViewToVertical = () => {
    setViewType('row view-row');
  };

  useEffect(() => (
    getArticlesHandler()
  ), []);

  return (
    <>
      <header className="page-header" data-pg-collapsed>
        <h1>Bookmarks </h1>
        <div className="btn-group btn-group-toggle float-right view-changer" data-toggle="buttons">
          <label className="btn active" onClick={changeViewToVertical}>
            <input type="radio" name="options" autoComplete="off" checked value="" />
            <i className="fas fa-th" />
          </label>
          <label className="btn" onClick={changeViewToHorizontal}>
            <input type="radio" name="options" autoComplete="off" value="list-view" />
            <i className="fas fa-list" />
          </label>
        </div>
      </header>
      {
        articles.articles.length !== 0
          ? (
            <>
              <div className={viewType}>
                {
                  // eslint-disable-next-line arrow-body-style
                  articles.articles.map((article, indx) => {
                    return (
                      <BookmarkItem
                        key={indx}
                        category={article.Article.articleCategory.category}
                        image={article.Article.image}
                        articleTitle={article.Article.title}
                        articleDescription={article.Article.description}
                        authorsName={article.Article.User.Profile.firstname}
                        authorsImage={article.Article.User.Profile.image}
                        readTime={article.Article.readTime}
                      />
                    );
                  })
                }
              </div>
              {/** pagination support to be implemented later */}
              <div>Pagination is here</div>
            </>
          ) : <div>
            <i className="fas fa-book-open" style={{ fontSize: '70px', color: '#66008c' }} />
            <p style={{ paddingTop: '30px' }}>*It looks like you have not bookmarked any article yet</p>
          </div>
      }
    </>
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getBookmarkLoading,
    articles: selectors.getBookmarkArticles,
  }
);

export default connect(mapStateToProps, { getBookmarkedArticles })(BookmarkContainer);
