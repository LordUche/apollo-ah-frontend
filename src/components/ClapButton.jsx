/* eslint-disable no-undef */
import React, { Component } from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { debounce } from 'lodash';
import { withRouter } from 'react-router-dom';
import clapIcon from '../images/clap.svg';
import * as articleSelector from '../selectors/singleArticleSelector';
import { getAuthId, getIsAuthenticated } from '../selectors/authSelector';
import { getToken } from '../selectors/profileSelector';
import { clapArticleRequest, fetchUserClaps } from '../actions/clapsAction';
import { getUserClaps } from '../selectors/clapsSelector';
import { CLAPS_LIMIT, CLAPS_REQUEST_TIMER } from '../utils/contants';

/**
 * A class representing clap button component
 *
 * @class ClapButton
 * @extends {Component}
 */
export class ClapButton extends Component {
  /**
   * Handles clap request (debounced)
   *
   * @returns {void}
   * @memberof ClapButton
   */
  makeClapRequest = debounce(() => {
    const { clapArticle, articleSlug: slug, token } = this.props;
    const { clapsCount: claps } = this.state;
    this.setState({ clapsCount: 0 });
    return clapArticle({ slug, claps, token });
  }, CLAPS_REQUEST_TIMER);

  static propTypes = {
    articleSlug: PropType.string.isRequired,
    articleAuthorId: PropType.number.isRequired,
    articleClaps: PropType.number.isRequired,
    clapArticle: PropType.func.isRequired,
    isLoggedIn: PropType.bool.isRequired,
    userId: PropType.number,
    history: PropType.object.isRequired,
    token: PropType.string.isRequired,
    userClaps: PropType.number.isRequired,
    loadUserClaps: PropType.func.isRequired,
  };

  static defaultProps = {
    userId: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      clapsLeft: CLAPS_LIMIT - props.userClaps,
      articleClaps: props.articleClaps,
      clapsCount: 0, // record clap button clicks
    };
  }

  componentDidMount = async () => {
    const {
      articleSlug: slug,
      isLoggedIn,
      userId,
      loadUserClaps
    } = this.props;
    if (isLoggedIn) {
      loadUserClaps({ slug, userId });
    }
  }

  /**
   * Makes an article clap
   *
   * @returns {void}
   * @memberof ClapButton
   */
  handleClapArticle = () => {
    const {
      isLoggedIn,
      userId,
      articleAuthorId,
      history,
    } = this.props;

    if (!isLoggedIn) { // redirect guest users to login
      history.push('/login');
      return false;
    }
    // Deny article's author
    if (articleAuthorId === userId) return false;
    const { articleClaps, clapsLeft, clapsCount } = this.state;
    if (clapsLeft >= 1) {
      this.setState({
        clapsLeft: clapsLeft - 1,
        articleClaps: articleClaps + 1,
        clapsCount: clapsCount + 1,
      });
      this.makeClapRequest();
    }
  }

  /**
   * React render function
   *
   * @returns {JSX.Element} DOM element
   * @memberof ClapButton
   */
  render() {
    const { articleClaps } = this.state;

    return (
      <span className="clap-icon">
        <button type="button" className="btn" onClick={this.handleClapArticle}>
          <img src={clapIcon} alt="" />
        </button>
        {articleClaps}
        {' '}
        claps
      </span>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  articleSlug: articleSelector.getArticleSlug,
  articleAuthorId: articleSelector.getArticleAuthorId,
  articleClaps: articleSelector.getArticleClaps,
  isLoggedIn: getIsAuthenticated,
  userId: getAuthId,
  token: getToken,
  userClaps: getUserClaps,
});

const mapDispatchToProps = {
  clapArticle: claps => clapArticleRequest(claps),
  loadUserClaps: payload => fetchUserClaps(payload),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(ClapButton));
