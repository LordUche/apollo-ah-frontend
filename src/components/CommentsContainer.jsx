/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';

import Comments from '../views/Comments';

/**
 * Container component for the Article comments view
 * @export
 * @class CommentsContainer
 * @extends {Component}
 */
export default class CommentsContainer extends Component {
  render() {
    return (
      <Comments />
    );
  }
}
