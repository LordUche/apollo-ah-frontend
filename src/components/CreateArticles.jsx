/* eslint-disable no-unused-vars */
/* eslint-disable arrow-parens */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import * as selectors from '../selectors/createArticleSelector';
import createArticleAction from '../actions/createArticles';
import CreateArticle from '../views/CreateArticle';
import CloudinaryWidget from './common/CloudinaryWidget';


/**
 * Form for creating a new article
 * @param {object} props Component props
 * @param {function} props.createArticleAction action creator function
 */
function CreateArticleContainer(props) {
  const [formData, setFormData] = useState({
    title: null,
    description: null,
    body: null,
    categoryId: null,
    tagList: [],
    image: null,
  });

  /** passing global state loading as props into variables */
  let { message, isLoading, token } = props;

  let widget;

  const updateInputHandler = (event) => {
    setFormData({ ...formData, [event.target.name]: event.target.value });
  };

  const updateBodyHandler = (value) => {
    setFormData({ ...formData, body: value });
  };

  const updateTagHandler = (tags) => {
    const listOfTags = [];
    tags.forEach(tag => {
      listOfTags.push(tag.text);
      setFormData({ ...formData, tagList: listOfTags });
    });
  };

  const submitHandler = (event) => {
    event.preventDefault();
     console.log('--->', formData);
     const requestData = {
       formData,
       token,
     };
    //props.createArticleAction(formData);
    props.createArticleAction(requestData);
  };

  const cloudinaryUpdate = (imageUrl) => {
    setFormData({ ...formData, image: imageUrl });
  };

  const imageUploadHandler = () => {
    // console.log('yo mama');
    // const { showLoader } = this.props;
    // showLoader(true);
    if (widget === undefined) {
      widget = new CloudinaryWidget(
        cloudinaryUpdate, (error) => {
        return (error);
      }, false);
    }
    widget.open();
  };

  console.log('--->', isLoading);
  console.log('==->', message);

  return (
    <CreateArticle
      updateInputHandler={updateInputHandler}
      updateBodyHandler={updateBodyHandler}
      updateTagHandler={updateTagHandler}
      imageUploadHandler={imageUploadHandler}
      submitHandler={submitHandler}
      valueText={formData.body}
    />
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getCreateArticleLoadingState,
    message: selectors.getCreateArticleMessageState,
    token: selectors.getUserToken,
  }
);

export default connect(mapStateToProps, { createArticleAction })(CreateArticleContainer);
