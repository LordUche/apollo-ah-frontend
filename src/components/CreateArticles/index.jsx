/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React from 'react';

function CreateArticle(props) {
  return (
    <section className="wrapper min-vh-100">
      <div className="container">
        <header className="page-header" data-pg-collapsed>
          <h1> New Article</h1>
        </header>
        <div className="card mb-5 mt-n5">
          <form role="form" className="form p-sm-4 p-3 form-borderless">
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="title" className="font-weight-bold">Title</label>
              <input type="text" className="form-control text-capitalize" id="title" name="title" required />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="description" className="font-weight-bold">Short Description</label>
              <textarea className="form-control textarea-sm" id="description" name="description" />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label htmlFor="description" className="font-weight-bold">Details</label>
              <textarea className="form-control" id="body" name="body" />
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label className="font-weight-bold" htmlFor="category">Category</label>
              <select className="custom-select" name="category" id="category">
                <option selected value="-1">-- Choose one --</option>
                <option value="3">Business</option>
                <option value="7">Entertainment</option>
                <option value="4">Fashion</option>
                <option value="6">Food</option>
                <option value="5">Health</option>
                <option value="8">History</option>
                <option value="1">Technology</option>
                <option value="2">Sports</option>
                <option value="9">Others</option>
              </select>
            </div>
            <div className="form-group mb-4" data-pg-collapsed>
              <label className="font-weight-bold" htmlFor="tags">Tags (optional)</label>
              <input type="text" className="form-control" id="tags" name="tags" data-role="tagsinput" />
            </div>
            <div className="form-group mb-4">
              <div className="custom-file">
                <input type="file" className="custom-file-input" id="custom-file1" />
                <label className="custom-file-label" htmlFor="custom-file1">Upload cover image</label>
              </div>
            </div>
            <button type="submit" className="btn btn-brand btn-plain mt-3">Publish article</button>
          </form>
        </div>
      </div>
    </section>
  );
}

export default CreateArticle;
