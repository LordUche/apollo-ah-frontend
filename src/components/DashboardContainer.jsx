import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getDashboard, getDashboardSummary } from '../selectors/dashboardSelector';
import { getProfile } from '../selectors/profileSelector';
import Dashboard from '../views/Dashboard';
import ProfileHeaderContainer from './ProfileHeaderContainer';
import { getDashboardArticles } from '../actions/dashboardAction';
import CommentLoader from '../views/CommentLoader';

/**
 * Container component for the Dashboard view
 * @export
 * @class DashboardContainer
 * @extends {Component}
 */
class DashboardContainer extends Component {
  async componentDidMount() {
    const { profile: { id } } = this.props;
    const { getDashboardArticles: fetchDashboardArticles } = this.props;
    await fetchDashboardArticles(id);
  }

  render() {
    const { dashboard, summary } = this.props;
    return (
      <>
        <ProfileHeaderContainer activePage="DASHBOARD" />
        {
          dashboard.loading
            ? <CommentLoader />
            : <Dashboard summary={summary} dashboard={dashboard} />
        }
      </>
    );
  }
}

DashboardContainer.propTypes = {
  dashboard: PropTypes.object.isRequired,
  summary: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
  getDashboardArticles: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector(
  {
    dashboard: getDashboard,
    summary: getDashboardSummary,
    profile: getProfile,
  }
);
const mapDispatchToProps = dispatch => ({
  getDashboardArticles: payload => dispatch(getDashboardArticles(payload)),
});
export default connect(mapStateToProps, mapDispatchToProps)(DashboardContainer);
