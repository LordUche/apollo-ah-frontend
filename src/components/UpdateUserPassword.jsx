/* eslint-disable react/no-access-state-in-setstate */
/* eslint-disable react/jsx-no-bind */
import React, { Component } from 'react';
import jwtDecode from 'jwt-decode';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import * as selectors from '../selectors/updatePasswordSelector';
import { updatePassword } from '../actions/updatePasswordAction';
import UpdatePasswordView from '../views/UpdatePassword';

/**
 * Form for resetting password
 * @class UpdateUserPassword
 * @extends {Component}
 * @param {object} props Component props
 * @param {function} props.updatePassword action creator function
 * @returns {JSX}
 */
export class UpdateUserPassword extends Component {
  // local state for handling email gotten from the input field
  state = {
    password: null,
    confirmPassword: null,
  }

  // this is a function to update the password and confirmPassword state
  updateInput = (event) => {
    this.setState({
      ...this.state,
      [event.target.name]: event.target.value
    });
  };

  // A function that triggers the post request action
  handleSubmit = async (event) => {
    event.preventDefault();
    const token = new URLSearchParams(window.location.search).get('token');
    const { email } = jwtDecode(token);
    const { password, confirmPassword } = this.state;
    const validateInput = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,}$/;
    if (!password.match(validateInput)) {
      toast.error('Password must contain uppercase and lowercase letters, a number and not be less than 6 characters');
    } else if (password === confirmPassword) {
      // calling the action creator to update the store
      const { updatePassword, message, history } = this.props;
      updatePassword(token, password, confirmPassword, email);
      if (message && message === 'Password reset successful') {
        toast.success('Password updated successfully !');
      }
      if (message && message !== 'Password reset successful') {
        toast.error('Your request to reset password has expired or check your internet connection.');
      }
      history.push('/login');
    } else {
      toast.error('Password do not match');
    }
  };

  /**
   * @memberof UpdateUserPassword
   * @returns {JSX} The JSX representation of the Reset Password component
  */
  render() {
    return (
      <UpdatePasswordView
        handleSubmit={this.handleSubmit}
        updateInput={this.updateInput}
      />
    );
  }
}

UpdateUserPassword.propTypes = {
  updatePassword: PropTypes.func,
  history: PropTypes.object,
  message: PropTypes.string,
};

UpdateUserPassword.defaultProps = {
  updatePassword: () => {},
};

const mapStateToProps = createStructuredSelector({
  message: selectors.getUpdatePasswordMessage,
});


export default connect(mapStateToProps, { updatePassword })(UpdateUserPassword);
