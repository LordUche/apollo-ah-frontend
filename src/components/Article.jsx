import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import ArticleBody from '../views/ArticleBody';
import CommentsContainer from './CommentsContainer';

/**
 *
 *
 * @class Article
 * @extends {React.Component}
 */
class Article extends React.Component {
  /**
   *
   * @static - Validating proptypes
   * @memberof Article
   */
  static propTypes = {
    match: PropTypes.object.isRequired,
  }

  /**
   * Creates an instance of Article.
   * @param {*} props - Props passed down from the BrowserRouter
   * @memberof Article
   */
  constructor(props) {
    super(props);
    this.state = {
      article: {},
      bookmarked: false,
    };
  }

  /**
   *
   * @todo - Move the fetch article functionality to redux store
   * @memberof Article
   */
  componentDidMount = async () => {
    const { match } = this.props;
    const { slug } = match.params;
    const result = await axios.get(`${process.env.API_BASE_URL}/articles/${slug}`);
    this.setState({ article: result.data.data });
  }

  /**
   * @todo - employ javascript debounce to avoid calling API on repeated clicks
   * @method bookmarkArticle - The bookmark article functionality
   * @memberof Article
   */
  bookmarkArticle = () => {
    const { bookmarked } = this.state;
    this.setState({ bookmarked: !bookmarked });
  };

  /**
   *
   *
   * @returns {JSX} The JSX representation of the Article component
   * @memberof Article
   */
  render() {
    const { article, bookmarked } = this.state;
    return (
      <div>
        <main className="main-body">
          <section className="min-vh-100">
            {article.image && (
              <div
                className="pg-empty-placeholder single-hero"
                style={{ backgroundImage: `url(${article.image})` }}
              />
            )}
            <div className="container">
              <div className="single-container">
                <ArticleBody
                  article={article}
                  bookmarkArticle={this.bookmarkArticle}
                  bookmarked={bookmarked}
                />
                <CommentsContainer />
              </div>
            </div>
            <div className="single-suggested-grp">
              <div className="container">
                <h4>Recommended for you</h4>
                <div className="row pg-empty-placeholder" />
              </div>
            </div>
          </section>
        </main>
      </div>
    );
  }
}

export default Article;
