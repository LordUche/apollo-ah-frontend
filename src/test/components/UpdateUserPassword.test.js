import React from 'react';
import { shallow } from 'enzyme';
import { UpdateUserPassword } from '../../components/UpdateUserPassword';

describe('<UpdateUserPassword Test Suite>', () => {
  const props = {
    message: 'Hello',
    isLoading: false,
    updatePassword: jest.fn(),
    updatePasswordMessage: 'Password updated successfully',
  };
  const event = {
    target: {
      name: 'password',
      value: 'mypassword'
    },
    preventDefault: jest.fn(),
  };
  it('It should render succesfully when user is not logged in', () => {
    const wrapper = shallow(<UpdateUserPassword {...props} />);
    expect(wrapper).toBeDefined();
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('should update password with users password', () => {
    const wrapper = shallow(<UpdateUserPassword {...props} />);
    wrapper.instance().updateInput(event);
    expect(wrapper.instance().state.password).toEqual('mypassword');
  });
});
