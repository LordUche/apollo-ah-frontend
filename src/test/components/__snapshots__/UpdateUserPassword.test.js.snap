// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateUserPassword Test Suite> It should render succesfully when user is not logged in 1`] = `
<UpdatePassword
  handleSubmit={[Function]}
  updateInput={[Function]}
/>
`;
